server/server/src/metrics.rs

```diff
@@ -0,0 +1,215 @@
+
+            if is_user_active(&metadatas).await? {
+                active_users += 1;
+            }
+
+            tokio::time::sleep(server_state.config.metrics.duration_between_user_metrics).await;
+        }
+
+        METRICS_STATISTICS
+            .total_documents
+            .set(total_documents as i64);
+        METRICS_STATISTICS.active_users.set(active_users);
+        METRICS_STATISTICS.total_document_bytes.set(total_bytes);
+
+        tokio::time::sleep(server_state.config.metrics.duration_between_metrics_refresh).await;
+    }
+}
+
+pub async fn get_all_public_keys(
+    server_state: &Arc<ServerState>,
+) -> Result<Vec<PublicKey>, ServerError<MetricsError>> {
+    let mut con = server_state.index_db_pool.get().await?;
+
+    let mut public_keys_k: AsyncIter<String> = con.scan_match(public_key("*")).await?;
+
+    let mut public_keys_keys = HashSet::new();
+    while let Some(item) = public_keys_k.next_item().await {
+        public_keys_keys.insert(item);
+        tokio::time::sleep(
+            server_state
+                .config
+                .metrics
+                .duration_between_getting_pub_key_key_metrics,
+        )
+        .await;
+    }
+
+    let mut public_keys: Vec<PublicKey> = vec![];
+
+    for key in public_keys_keys {
+        let public_key = con.maybe_json_get(&key).await?.ok_or_else(|| {
+            internal!(
+                "Cannot retrieve public_key despite having a valid key: {:?}",
+                key
+            )
+        })?;
+
+        public_keys.push(public_key);
+        tokio::time::sleep(
+            server_state
+                .config
+                .metrics
+                .duration_between_getting_pub_key_metrics,
+        )
+        .await;
+    }
+
+    Ok(public_keys)
+}
+
+pub async fn get_users_owned_files(
+    con: &mut deadpool_redis::Connection,
+    public_key: &PublicKey,
+) -> Result<Vec<Uuid>, ServerError<MetricsError>> {
+    con.maybe_json_get(keys::owned_files(public_key))
+        .await?
+        .ok_or_else(|| {
+            internal!(
+                "Cannot retrieve owned_files despite having a valid public_key: {:?}",
+                public_key
+            )
+        })
+}
+
+pub async fn get_user_file_metadatas(
+    con: &mut deadpool_redis::Connection,
+    ids: &[Uuid],
+) -> Result<Vec<EncryptedFileMetadata>, ServerError<MetricsError>> {
+    let mut metadatas = vec![];
+
+    for id in ids {
+        let metadata: EncryptedFileMetadata =
+            con.maybe_json_get(keys::file(*id)).await?.ok_or_else(|| {
+                internal!("Cannot retrieve encrypted file metadata despite having a valid id.")
+            })?;
+
+        metadatas.push(metadata)
+    }
+
+    Ok(metadatas)
+}
+
+pub async fn calculate_total_document_bytes(
+    con: &mut deadpool_redis::Connection,
+    metadatas: &[EncryptedFileMetadata],
+) -> Result<i64, ServerError<MetricsError>> {
+    let mut total_size: u64 = 0;
+
+    for metadata in metadatas {
```

Comment:
doing it this way is fine... but you could do:
`for metadata in metadatas.filter_not_deleted()` which will parse the tree and remove everything that's not deleted. Maybe you want to do that in `get_user_file_metadatas`, maybe not because a recently deleted file indicates user activity.

Reply:
I think filtering deleted files isn't really necessary since they don't really cause any problems in the code. Right? Wouldn't filtering instead just make metrics run a bit longer.

Reply:
you actually do need to do it this way because right now we're counting deleted files towards our total file count.

but you need to do it in a way that deletion still counts towards active users. In the main loop `total_documents += ids.len();` is wrong because it will include deleted files. 

IMO in your main loop you should first calculate if a user is active and then set `metadatas = metadatas.filter_deleted()` and then the rest of your functions can remain unchanged.

Reply:
Okay that makes sense.